text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,533 @@
+       has a limit of 500 MB.
+    3. PyPi has an unofficially documented limit of 60 MB.
+2. A large download size might be problematic for customers with weak internet
+   connections or limited bandwidth.
+3. Customers who download and install `aws-cdk-lib` on every CI/CD pipeline run
+   may be frustrated to spend resources frequently downloading a large package.
+4. AWS Lambda limits the size of your function code to 250 MB.
+5. The AWS Lambda limit for all uploaded assets for a Function is 75 GB. If a
+   customer is using Lambda Versions, and wants to maintain a Version for every
+   version they have published, they might end up hitting this limit because of
+   the large size of `aws-cdk-lib`.
+6. The developer experience of using `aws-cdk-lib` in an IDE might be slower in
+   some cases.
+7. The CDK project is currently blocked from including more ‘large
+   dependencies’, such as multiple version of kubectl, multiple version of the
+   AWS CLI, and dedicated versions of the AWS SDK for use in custom resources.
+   This is a decision we’ve made to avoid increasing the size of the package
+   further until a more scalable solution for these kinds of dependencies is
+   implemented.
+
+Most of these above issues have workarounds. For running CDK apps in Lambda, you
+can use a Docker image for your Lambda code, and the limit becomes 10GB. For
+package manager size limits, we can request
+[increases](https://github.com/pypa/pypi-support/issues/1642), and those have
+historically been approved. However, publishing a large package is a bad
+practice in the open-source software ecosystem. Even if a customer has no
+technical limitations for adding a 234 MB dependency to their package, it is a
+red flag to them when considering using `aws-cdk-lib`. We should be
+customer-obsessed, and solve the problem of `aws-cdk-lib` being a huge
+dependency, before customers start complaining about it.
+
+### Why is the package so large today?
+
+There are some large assets contributing to the size of `aws-cdk-lib`, like
+.jsii files, and zip files of dependencies used in custom resources. The library
+also contains a huge volume of source code now that `aws-cdk-lib` combines 231
+(and increasing) CDK modules. The table below breaks down the percentage that
+each category of files contributes to the size of aws-cdk-lib.
+
+Note: The percentages do not add up to 100. Some files, e.g. various .json
+files, are excluded and contribute very little to the size.
+
+| Category                       | Percentage of aws-cdk-lib size |
+| ------------------------------ | ------------------------------ |
+| .jsii files                    | 41.64%                         |
+| Source Maps (.js.map)          | 21.93%                         |
+| Lambda Layer zip files         | 15.48%                         |
+| Javascript code (.js)          | 9.17%                          |
```

Comment:
Does this include `.warnings.jsii.js` files? There is some room for improvement there: if there is no warning to print for a given class, it will still generate a corresponding function, taking up unnecessary space. We can change the warnings generation in JSII to not generate these functions and not call them from the compiled .js files. But, of course, it depends on how much this particular issue contributes to the whole. Might be worth it getting some data around this.